src/Store.ts

```diff
@@ -0,0 +1,392 @@
+	 * so their is a clear delineation between new objects and restored objects.
+	 *
+	 * @param object The raw object with the properties that need to be defined on the new
+	 * model instance
+	 * @param mutateAllowed This indicates if restore is allowed to mutate the original object
+	 * (by setting its __proto__). If this isn't true, than the restore should
+	 * copy the object to a new object with the correct type.
+	 * @returns An instance of the store model, with all the properties that were defined
+	 * on object. This may or may not be the same object that was passed in.
+	 */
+	protected _restore(object: Hash<any>, mutateAllowed: boolean) {
+		const Model = this.Model;
+		if (Model && object) {
+			const prototype = Model.prototype;
+			const restore = prototype._restore;
+			if (restore) {
+				// the prototype provides its own restore method
+				object = restore.call(object, Model, mutateAllowed);
+			} else if (hasProto && mutateAllowed) {
+				// the fast easy way
+				// http://jsperf.com/setting-the-prototype
+				object['__proto__'] = prototype;
+			} else {
+				// create a new object with the correct prototype
+				object = <Hash<any>> lang.create(prototype, object);
+			}
+		}
+		return object;
+	}
+
+	/**
+	 * Sets an object's identity description:
+	 * This method sets an object's identity and is useful to override to support
+	 * multi-key identities and object's whose properties are not stored directly on the object.
+	 *
+	 * @param object The target object
+	 * @param identityArg The argument used to set the identity
+	 */
+	protected _setIdentity(object: { [ index: string ]: any; set: (...args: any[]) => any}, identityArg: string | number) {
+		if (object.set) {
+			object.set(this.idProperty, identityArg);
+		} else {
+			object[this.idProperty] = identityArg;
+		}
+	}
+
+	/**
+	 * This creates a new instance from the store's model.
+	 * @param properties The properties that are passed to the model constructor to
+	 * be copied onto the new instance. Note, that should only be called
+	 * when new objects are being created, not when existing objects
+	 * are being restored from storage.
+	 */
+	create(properties: {}) {
+		return new this.Model(properties);
+	}
+
+	/**
+	 * Retrieves a range of objects from the collection, returning a promise to an array.
+	 *
+	 * @param args Contains the starting index of objects to return (0-indexed)
+	 * and the exclusive end of objects to return
+	 * @return A FetchPromise that will resolve with the results of the fetch
+	 */
+	fetchRange(args: dstore.FetchRangeArgs):dstore.FetchPromise<T> {
+		throw new Error("This method is abstract");
+	}
+
+	filter<T>(...args: any[]): dstore.Collection<T> {
+		return QueryMethod(<QueryMethodArgs<T>> {
+			type: 'filter',
+			normalizeArguments: function (filter) {
+				const Filter = this.Filter;
+				if (filter instanceof Filter) {
+					return [ filter ];
+				}
+				return [ new Filter(filter) ];
+			}
+		}).apply(this, args);
```

Comment:
It actually is meaningful in this context. This used to be `filter: new QueryMethod`, so that `filter` would be assigned the result of calling `QueryMethod`, and when it was executed via `store.filter`, `this` inside the function would refer to this store(and it is expected to behave as such). Now if this is going to be a method it can't be assigned to `QueryMethod(...)` so the `apply(this` is needed to provide context.

I'm wondering now though what would be more efficient, executing QueryMethod once each time a store is insantiated or executing it each time `store.filter` is called. I think it's nicer to have this as a method but than a property but I don't know if that justifies the extra function calls.